pkg/trace/writer/sender.go

```diff
@@ -155,14 +157,34 @@ func (s *queuableSender) Start() {
 // Run executes the queuableSender main logic synchronously.
 func (s *queuableSender) Run() {
 	defer close(s.exit)
+	sema := make(chan struct{}, s.conf.MaxConnections)
 
 	for {
 		select {
 		case payload := <-s.in:
-			if stats, err := s.sendOrQueue(payload); err != nil {
-				log.Debugf("Error while sending or queueing payload. err=%v", err)
-				s.notifyError(payload, err, stats)
+			if payload == nil {
+				continue
+			}
+			s.mu.RLock()
+			queuing := s.queuedPayloads.Len() > 0
+			s.mu.RUnlock()
+			if queuing {
+				if err := s.enqueue(payload); err != nil {
+					log.Debugf("Error while queueing payload: %v", err)
+					s.notifyError(payload, err)
+				}
+				continue
 			}
+			sema <- struct{}{}
+			s.wg.Add(1)
+			go func() {
+				defer s.wg.Done()
+				if err := s.trySend(payload); err != nil {
```

Comment:
Nit: since `trySend` already notifies on success and retry, may simplify this code to move the `notifyError` responsibility into `trySend` as well.